@xen-orchestra/proxy/src/main/index.js

```diff
@@ -1,13 +1,43 @@
-import { camelCase } from 'lodash'
+import camelCase from 'lodash/camelCase'
 
 import mixins from './mixins'
 
+const { defineProperties, defineProperty, keys } = Object
+const noop = Function.prototype
+
+const MIXIN_CYCLIC_DESCRIPTOR = {
+  configurable: true,
+  get() {
+    throw new Error('cyclic dependency')
+  },
+}
+
 export default class App {
   constructor(opts) {
-    Object.keys(mixins).forEach(mixin => {
-      Object.defineProperty(this, camelCase(mixin), {
-        value: new mixins[mixin](this, opts),
-      })
+    // add lazy property for each of the mixin, this allow mixins to depend on
+    // one another without any special ordering
+    const descriptors = {}
+    keys(mixins).forEach(name => {
+      const Mixin = mixins[name]
+      name = camelCase(name)
+
+      descriptors[name] = {
+        configurable: true,
+        get: () => {
+          defineProperty(this, name, MIXIN_CYCLIC_DESCRIPTOR)
```

Comment:
It's necessary?

Reply:
Yes and no, without it cyclic dependencies (i.e. two mixins using one another in their constructors) will trigger a `Maximum call stack size exceeded` error.

This just make the debugging easier by failing fast and clearly.